core/tx_list.go

```diff
@@ -270,23 +276,25 @@ func (l *txList) Forward(threshold uint64) types.Transactions {
 	return l.txs.Forward(threshold)
 }
 
-// Filter removes all transactions from the list with a cost higher than the
-// provided threshold. Every removed transaction is returned for any post-removal
-// maintenance. Strict-mode invalidated transactions are also returned.
+// Filter removes all transactions from the list with a cost or gas limit higher
+// than the provided thresholds. Every removed transaction is returned for any
+// post-removal maintenance. Strict-mode invalidated transactions are also
+// returned.
 //
-// This method uses the cached costcap to quickly decide if there's even a point
-// in calculating all the costs or if the balance covers all. If the threshold is
-// lower than the costcap, the costcap will be reset to a new high after removing
-// expensive the too transactions.
-func (l *txList) Filter(threshold *big.Int) (types.Transactions, types.Transactions) {
+// This method uses the cached costcap and gascap to quickly decide if there's even
+// a point in calculating all the costs or if the balance covers all. If the threshold
+// is lower than the costgas cap, the caps will be reset to a new high after removing
+// the newly invalidated transactions.
+func (l *txList) Filter(costLimit, gasLimit *big.Int) (types.Transactions, types.Transactions) {
 	// If all transactions are below the threshold, short circuit
-	if l.costcap.Cmp(threshold) <= 0 {
+	if l.costcap.Cmp(costLimit) <= 0 && l.gascap.Cmp(gasLimit) <= 0 {
 		return nil, nil
 	}
-	l.costcap = new(big.Int).Set(threshold) // Lower the cap to the threshold
+	l.costcap = new(big.Int).Set(costLimit) // Lower the caps to the thresholds
```

Comment:
A transaction list is never manipulated concurrently. It resides in the transaction pool which ensures serialized access for everything.